src/main/scala/edu/ie3/simona/model/grid/GridModel.scala

```diff
@@ -592,6 +634,133 @@ case object GridModel {
     gridModel
   }
 
+  /** Build business models for transformer control groups
+    *
+    * @param config
+    *   List of configs for control groups
+    * @param measurementUnitInput
+    *   Set of [[MeasurementUnitInput]] s
+    * @return
+    *   A set of control group business models
+    */
+  private def buildTransformerControlGroups(
+      config: List[SimonaConfig.TransformerControlGroup],
+      measurementUnitInput: java.util.Set[MeasurementUnitInput]
+  ): Set[ControlGroupModel] = config.map {
+    case TransformerControlGroup(measurements, _, vMax, vMin) =>
+      buildTransformerControlGroupModel(
+        measurementUnitInput,
+        measurements,
+        vMax,
+        vMin
+      )
+  }.toSet
+
+  /** Build a single control group model. Currently, only limit violation
+    * prevention logic is captured: The nodal regulation need is equal to the
+    * voltage change needed to comply with the given thresholds
+    *
+    * @param measurementUnitInput
+    *   Collection of all known [[MeasurementUnitInput]] s
+    * @param measurementConfigs
+    *   Collection of all uuids, denoting which of the [[MeasurementUnitInput]]
+    *   s does belong to this control group
+    * @param vMax
+    *   Upper permissible voltage magnitude
+    * @param vMin
+    *   Lower permissible voltage magnitude
+    * @return
+    *   A [[ControlGroupModel]]
+    */
+  private def buildTransformerControlGroupModel(
+      measurementUnitInput: java.util.Set[MeasurementUnitInput],
+      measurementConfigs: List[String],
+      vMax: Double,
+      vMin: Double
+  ): ControlGroupModel = {
+    val nodeUuids =
+      determineNodeUuids(measurementUnitInput, measurementConfigs)
+    buildTransformerControlModels(nodeUuids, vMax, vMin)
+  }
+
+  /** Determine the uuids of the nodes to control
+    *
+    * @param measurementUnitInput
+    *   Collection of all known [[MeasurementUnitInput]] s
+    * @param measurementConfigs
+    *   Collection of all uuids, denoting which of the [[MeasurementUnitInput]]
+    *   s does belong to this control group
+    * @return
+    *   A set of relevant nodal uuids
+    */
+  private def determineNodeUuids(
+      measurementUnitInput: java.util.Set[MeasurementUnitInput],
+      measurementConfigs: List[String]
+  ): Set[UUID] = Set.from(
+    measurementUnitInput.asScala
+      .filter(input =>
+        measurementConfigs.contains(input.getUuid.toString) && input.getVMag
```

Comment:
This is something that we maybe should test as well during config fail fast for GridAgent: Do the measurement units that are assigned for transformer control groups measure voltage magnitude? If this is not the case, the device is filtered out here and this can lead to very confusing results, I can imagine

Reply:
Good idea. I moved the FailFast of GridAgent into it's own class. The solution we found together changed that way, that there is a check if the transformer belongs to the subgrid before checking the measurements. Please have a look if it make sense. @sebastian-peter